design-documents/services-orchestration.md

```diff
@@ -0,0 +1,254 @@
+So, by using native Magento XML config & merger we can declare or extend scenario in way we want.
+In the example above payment failure handling step was added by SalesPayment module, which knows about `placeOrder` scenario.
+And extends it with a new logic.
+```xml
+        <step xsi:type="Task" name="PlacePaymentForOrder" next="PersistOrder">
+            <catch exceptoin="Magento\Payment\Gateway\Command\CommandException" next="PaymentFailureHandler" />
+        </step>
+        <step xsi:type="Task"
+            name="PaymentFailureHandler"
+            inputPath="params/order"
+            exit="true"
+        />
+```
+
+### Replaceability
+With assumption that a service is replacable, everything that exists inside of a service implementation boundary is mutable.
+We should not expect that code inside of services will trigger the same set of events in a future(because service can be replaced).
+The beauty of scenarios in a fact that you can replace a scenario step, if you need to replace the step only.
+Or you can redefine scenario if you need this. Both of cases are valid and depend on business requirements you have.
+* By changing only step a developer works only with this step(s) developer introduced.
+* Step does not have a strict interface of arguments that will be passed. This list vary depends on the step configuration.
+    * Changing step, route to a function with an interface that is different from core one:
+```xml
+<step xsi:type="Task" name="PlacePaymentForOrder" next="PersistOrder"
+      inputPath="params/order"
+      outputPath="params/order"
+      routine="Magento:MySales:OrderManagement:placeOrder"
+/>
+```
+
+*
+   * Extending the existing one scenario with more steps
+```xml
+<step xsi:type="Task" name="PlacePaymentForOrder" next="DoSomethingPersistOrder"/>
+<step xsi:type="Task" name="DoSomethingPersistOrder" next="PersistOrder"/>
+```
+* Anyway if your code has the same interface with existing routine you can substitute routine only.
+* If a developer replaced the whole scenario this is BiC change. But, quite visible and easy catchable even without analyzing PHP code.
+All you need you can find in configuration (I did not investigate this question but pretty sure that it is possible to build visualization for scenario steps).
+
+### Events and Observers
+
+Step execution can be treated as an event in the system. So we can declare an observer on it.
+```xml
+<step xsi:type="Observer" name="SendEmail" after="PersistOrder"
+    next="PersistOrder"
+    inputPath="params/order"
+    routine="Magento:MySales:OrderManagement:sendEmail"
+/>
+
+```
+Observer can not return values into `outputPath` or start a new execution branch.
+As a result, observer execution can be delayed or even processed async in background.
+Observer can be declared `before` of `after` the step.
+
+![Figure 2](services-orchestration/events.png)
+
+## Sync and Async adapters
+Ideally, all scenarios should not be run synchronously.
```

Comment:
I think some scenarios can be only synchronous. And some scenarios (add to cart for instance) would be best to keep synchronous. What store front scenarios except modifying when totals being calculated and placing order would you make asynchronous?

Reply:
Scenario declaration does not depend on the manner how the scenario could be executed. Sync or async mode will be chosen by a corresponding scenario invoker (adapter).